### PR TITLE
Implementation of Portable TeeUniverse Maps

### PR DESCRIPTION
**Pull request by necropotame**
DDNet Testing Server: 137.74.156.196:10004

This commit modifies DDNet server to make it able to read Portable TeeUniverse Maps (PTU maps).
PTU maps are pure vanilla maps, but that contains hidden layers (clipped) containing animated zones (death zones, freeze zones, ...) and animated entities (laser, armor, ...).

Adventages for DDNet:
- Moving zones made of quads (currently: freeze, unfreeze, deep, undeep and death)
- Moving pickups (currently: all vanilla pickups)
- No problem of collision between tiles because freeze zones use a dedicated layer. In fact, any layer called "ddFreeze" in the group "#Zones" will be interpreted as freeze zone (quads or tiles). It means that you can use as many layers as you want is you like to separate things. The Index 1 is for freeze, 2 for unfreeze, 3 for deep and 4 for undeep. Indices for tiles are stored as usual, and indices for quads are stored in the field "Color TO".
- It can be extended without changing the format of the map. Just add, using the editor, a new layer in "#Zones" and you can start to use it in the server. Nothing is hard codded or specific to DDNet.
- This map format is designed to be completely editable using the vanilla editor and a kind of standard for mods. I'm not saying that DDNet should switch to it, but be able to read it and use it seems interesting.

### CODE REVIEW

**Comment by heinrich5991** [comment]
Sounds cool. Could you maybe run a test server with DDNet tiles? :)

**Comment by necropotame** [comment]
Sorry, wrong server address -> 137.74.156.196:10004

**Comment by heinrich5991** [comment]
What we should definitely think about is the sensitivity. Seems like you have to be halfway in for it to trigger right now.

**Comment by necropotame** [comment]
I've used the same collision function that for tiles: the center of the tee. It's up to you, DDNet players, to decide what is the best between precision and coherence with the actual system based on tiles :)

File: src/game/server/entities/character.cpp [hunk 1]
```diff
@@ -1933,6 +1946,23 @@ void CCharacter::HandleTiles(int Index)
 	}
 }
 
+void CCharacter::HandleZones()
+{
+	int FreezeIndex = GameServer()->Collision()->GetZoneValueAt(GameServer()->m_ZoneHandle_DDFreeze, m_Pos.x, m_Pos.y);
```
Comment (necropotame) [inline src/game/server/entities/character.cpp]: The precision is set here. I'm checking only m_Pos. See L. 1243 for a more precise check (the death zone case).

**Comment by necropotame** [comment]
So, since the gametick is merged, this PR can now work properly. Someone can tell me what to improve here, what to remove, etc...? I know that entity layer, as well as the editor, should support animated zones, but I want to get a signal before taking the time to implement it. @def- ? @east ?

**Comment by heinrich5991** [comment]
I''d like to go ahead with this. What are othere people's thoughts on this?

@necropotame How do I create such a map? Should I use the TeeUniverse editor?

**Comment by necropotame** [comment]
TeeUniverse doesn't support animations ATM ( but this will be done for the next release), however, you can create polygonal static zones. You can read this page (sorry for the English, I've written it without read it again) to know more about the specifications and how to create PTU maps: https://teeuniverse.net/doc/specification/ptum/

**Comment by necropotame** [comment]
Please create a dedicated branch so we can work on it without worrying about releases. We need to discuss how to support more entities, how to render the entity overlay and how to improved the experience of the editor.

**Comment by heinrich5991** [comment]
Merged into the branch 'teeuniverse'.

**Comment by heinrich5991** [comment]
We can work on from there.

**Comment by ZombieToad** [comment]
revive this please

**Pull request closed without merging**